File: grind75/Dec30/39.go
func combinationSum(candidates []int, target int) [][]int {
    ret := [][]int{}
    sort.Ints(candidates)
    n := len(candidates)
    var dfs func([]int, int, int)
    dfs = func(cur []int, idx int, sum int) {
        if sum == target {
            ret = append(ret, cur)
            return
        }
        for i := idx; i < n; i++ {
            if sum+candidates[i] > target {
                break
            }
            tmp := append([]int{}, cur...)
            tmp = append(tmp, candidates[i])
            dfs(tmp, i, sum+candidates[i])
        }
    }
    dfs([]int{}, 0, 0)
    return ret
}

File: grind75/Dec30/39.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> 
List[List[int]]:
        result = []
        def dfs(prefix, current_sum, current_i):
            if current_sum == target:
                result.append(prefix)
                return
            for i in range(current_i, len(candidates)):
                if current_sum + candidates[i] <= target:
                    dfs(prefix + [candidates[i]], current_sum + 
candidates[i], i)
        
        dfs([], 0, 0)
        return result
